### ISSUE TITLE
Nokogiri doesn't add -liconv to LDFLAGS and mistakenly reports a failure to find libxml when building libxslt

### ISSUE THREAD

**Issue by mojca** [sparklemotion/nokogiri#1423]
On @zenspider's request I'm summarizing my problems with building nokogiri (which might or might not be related to the initial report in #1334; it also looks to me as if some people commenting on #1206 might have experienced the same problem).

I'm using 10.7.

~~~~~
> which ruby
/Users/me/.rbenv/shims/ruby
> ruby --version
ruby 2.2.3p173 (2015-08-18 revision 51636) [x86_64-darwin11.0]
~~~~~

I first installed rbenv and then

~~~~~
rbenv install 2.2.3
rbenv global 2.2.3
# this is when problems with nokogiri popped up
gem install rails -v 4.2.4
~~~~~

I tried building both the released version as well as the master and both fail with:
~~~~~
checking for main() in -llzma... no
checking for xmlParseDoc() in libxml/parser.h... no
checking for xmlParseDoc() in -lxml2... no
checking for xmlParseDoc() in -llibxml2... no
-----
libxml2 is missing.  Please locate mkmf.log to investigate how it is failing.
~~~~~

However that message is misleading.

The file `~/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/extensions/x86_64-darwin-11/2.2.0-static/nokogiri-1.6.7.rc4/mkmf.log` says that the failure to find libxml

~~~~~
have_func: checking for xmlParseDoc() in libxml/parser.h... -------------------- no
~~~~~

is because of:

~~~~~
Undefined symbols for architecture x86_64:
  "_libiconv", referenced from:
      _xmlIconvWrapper in libxml2.a(encoding.o)
  "_libiconv_close", referenced from:
      _xmlFindCharEncodingHandler in libxml2.a(encoding.o)
      _xmlCharEncCloseFunc in libxml2.a(encoding.o)
  "_libiconv_open", referenced from:
      _xmlFindCharEncodingHandler in libxml2.a(encoding.o)
ld: symbol(s) not found for architecture x86_64
~~~~~

while the beginning of `gem_make.out` reports that iconv has been successfully found:

~~~~~
checking for iconv.h... yes
checking for gzdopen() in -lz... yes
checking for iconv... yes
~~~~~

One particularly suspicious part of the log is the following:

~~~~~
have_iconv?: checking for iconv... -------------------- yes

"clang -o conftest -I/Users/me/.rbenv/versions/2.2.3/include/ruby-2.2.0/x86_64-darwin11.0 -I/Users/me/.rbenv/versions/2.2.3/include/ruby-2.2.0/ruby/backward -I/Users/me/.rbenv/versions/2.2.3/include/ruby-2.2.0 -I. -I/Users/me/.rbenv/versions/2.2.3/include  -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT    -O3 -Wno-error=shorten-64-to-32  -pipe  conftest.c  -L. -L/Users/me/.rbenv/versions/2.2.3/lib -L. -L/Users/me/.rbenv/versions/2.2.3/lib  -fstack-protector -L/usr/local/lib     -lruby-static -framework CoreFoundation  -lpthread -ldl -lobjc  "
Undefined symbols for architecture x86_64:
  "_iconv", referenced from:
      _main in conftest-iiEMBa.o
  "_iconv_open", referenced from:
      _main in conftest-iiEMBa.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: #include <stdlib.h>
 4: #include <iconv.h>
 5: 
 6: int main(void)
 7: {
 8:     iconv_t cd = iconv_open("", "");
 9:     iconv(cd, NULL, NULL, NULL, NULL);
10:     return EXIT_SUCCESS;
11: }
/* end */
~~~~~

There is no `-liconv` in the command that calls `clang`. I don't understand why the "iconv detection" succeeds in the first place. But calling `clang` manually on the same source file without `-liconv` fails for me as well. And if I add `-liconv` manually, the compilation succeeds. (But I don't know where that flag should be added in the configuration files, so I don't know how to fix the problem.)

I tried compiling a minimal example:

~~~~~
#include <stdlib.h>
#include <iconv.h>

int main(void)
{
    iconv_t cd = iconv_open("", "");
    iconv(cd, NULL, NULL, NULL, NULL);
    return 0;
}
~~~~~

which certainly doesn't work unless I add `-liconv`:

~~~~~
> clang test.c 
Undefined symbols for architecture x86_64:
  "_iconv", referenced from:
      _main in conftest-CTmtzg.o
  "_iconv_open", referenced from:
      _main in conftest-CTmtzg.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
> clang test.c -liconv 
> [OK]
~~~~~

I workaround to make nokogiri work is to link against an external `libxml`/`libxslt`, but this is something that I want to avoid.

First of all I guess that after adding `-liconv` the compilation of `libxslt` would succeed. But it would probably be much easier if nokogiri would simply build its own version of `libiconv` in the same way as it builds the other two libraries. The extra cost would be negligible.

(Just for completeness: I have MacPorts installed, but I didn't see any traces of `nokogiri` actually trying to use any files from MacPorts and without providing any explicit environmental variables or flags, I also don't understand why `nokogiri` would even attempt to use anything from MacPorts.)

My questions:
  * Can someone please help me figure out how to add `-liconv` to the above mentioned invocation of clang that tests for `libiconv`?
  * I need explanation about why the search for `libiconv` reports success when it actually fails.
  * Can someone please advise me about the proper way to compile the above mentioned minimal example in C? If it indeed compiles for some without the extra `-liconv` flag, this is something that should be pretty straightforward to test.

**Comment by zenspider** [comment]
Thanks! Additional questions:

* What version of OSX? (`sw_vers`)
* Do you have homebrew as well as macports?
* What `lib*` entries do you have from either homebrew or macports?
* What compiler & version was used to build your ruby?

**Comment by mojca** [comment]
- 
`````
> sw_vers
ProductName:	Mac OS X
ProductVersion:	10.7.5
BuildVersion:	11G63b
`````
- No. I don't have HomeBrew as it could lead to problems when mixed with another package manager. But I installed `rbenv` via MacPorts. I "followed" instructions on https://gorails.com/setup/osx/ (without actually fetching Homebrew) and started drafting https://trac.macports.org/wiki/howto/RubyOnRails until I was stuck at nokogiri.
- Do you mean from `libiconv`? (I can list others, but the list is long.)
`````
/opt/local/lib/libiconv.2.dylib
/opt/local/lib/libiconv.a
/opt/local/lib/libiconv.dylib -> libiconv.2.dylib
/opt/local/lib/libiconv.la
`````
- I cannot be 100% sure. But I used `rbenv install 2.2.3`. I tried to repeat that now with `rbenv install 2.3.0-dev` and it seems that it uses `cc`:
`````
> which cc
/usr/bin/cc
> ll /usr/bin/cc 
lrwxr-xr-x  1 root  wheel  5 21 jun  2013 /usr/bin/cc -> clang
> cc --version
Apple LLVM version 4.2 (clang-425.0.28) (based on LLVM 3.2svn)
Target: x86_64-apple-darwin11.4.2
Thread model: posix
`````

**Comment by jesperronn** [comment]
Anybody who knows a workaround for this problem? Are there flags I can set or system libraries I can use? If so, which commands could I use? 

I ran into this exact same problem (but I'm using homebrew)

**Comment by jesperronn** [comment]
Update: I was able to bypass the error by updating to newest prerelease: 

Added to my Gemfile:

```
gem 'nokogiri', '~>1.6.8.rc3'
```

**Comment by mojca** [comment]
Trying to install nokogiri 1.6.8.rc3 just lead to a different error for me, already while building `libxml2-2.9.3`:
`````
  CC       xmlIO.lo
cc1: warning: -Wuninitialized is not supported without -O
xmlIO.c: In function ‘xmlXzfileClose’:
xmlIO.c:1450: error: ‘LZMA_OK’ undeclared (first use in this function)
xmlIO.c:1450: error: (Each undeclared identifier is reported only once
xmlIO.c:1450: error: for each function it appears in.)
make[2]: *** [xmlIO.lo] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
========================================================================
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.
`````
but I'm aware that this is a different issue that has to be reported elsewhere. It is strange though that I removed all traces of `/opt/local` from the `PATH` variable, but `/Users/me/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/nokogiri-1.6.8.rc3/ext/nokogiri/tmp/x86_64-apple-darwin11.4.2/ports/libxml2/2.9.3/libxml2-2.9.3/config.log` still lists `/opt/local/bin`. And some suboptimal logic in libxml2 seems to be a problem in that case.

But I still wonder: how can I avoid `/opt/local/bin` from sneaking into configuration?

**Comment by mojca** [comment]
I'm sorry, I take my words back. After some further diagnoses I realized that it is most probably impossible to remove  base path of `rbenv` from `SHELL PATH`, so I [fixed](https://trac.macports.org/changeset/147496/) the packing of `rbenv` in MacPorts to put all the executables inside a "hidden" directory and only made a symlink from `/opt/local/bin/rbenv` to `/opt/local/libexec/rbenv/rbenv`. I still believe that `libxml2`'s configuration is somewhat "broken" ([link](https://bugzilla.gnome.org/show_bug.cgi?id=764657)), but now at least it no longer calls `/opt/local/bin/pkg-config`, doesn't find `liblzma` and thus doesn't declare `HAVE_LIBLZMA=1` that would lead to a build failure.

I made sure I was using the latest version of `rbenv` and the latest version of `ruby` and installation of 1.6.8.rc3 now succeeded (while 1.6.7.2 still breaks "as expected").

I would still be interested to know which commit fixed that, but I'm very happy this was finally resolved.
